### PR TITLE
std.CompTimeStringMap*: support empty initialization list

### LINKED ISSUE #18212: `std.ComptimeStringMap` gives compile error when giving an empty list
### Zig Version

0.12.0-dev.1800+559e216f3

### Steps to Reproduce and Observed Behavior

When compiling the following code,
```zig
const std = @import("std");
pub fn main() void {
    const dict = std.ComptimeStringMap(i32, .{});
    // const dict = std.ComptimeStringMap(i32, .{.{"nothing", 0}});
    std.debug.print("{?}\n", .{dict.get("anything")});
}
```

it gives the following compile error:
```
/home/fsp/.zvm/master/lib/std/comptime_string_map.zig:74:35: error: indexing into empty array is not allowed
        const min_len = sorted_kvs[0].key.len;
                        ~~~~~~~~~~^~~
/home/fsp/.zvm/master/lib/std/comptime_string_map.zig:14:36: note: called from here
    return ComptimeStringMapWithEql(V, kvs_list, defaultEql);
           ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
1.zig:3:39: note: called from here
    const dict = std.ComptimeStringMap(i32, .{});
                 ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
referenced by:
    callMain: /home/fsp/.zvm/master/lib/std/start.zig:575:17
    initEventLoopAndCallMain: /home/fsp/.zvm/master/lib/std/start.zig:519:34
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

I believe the root cause is that `ComptimeStringMapWithEql` does not handle the special case that the list is empty.

### Expected Behavior

When providing an empty list, `std.ComptimeStringMap` shall behave as an empty dict, instead of generating compile error.

### PR DESCRIPTION
**Pull request by rootbeer**
Add tests for empty initialization, and some more corner cases (empty key, very long key, duplicate keys).

Fixes #18212

### CODE REVIEW

File: lib/std/comptime_string_map.zig [hunk 1]
```diff
@@ -43,6 +43,7 @@ pub fn ComptimeStringMapWithEql(
     comptime kvs_list: anytype,
     comptime eql: fn (a: []const u8, b: []const u8) bool,
 ) type {
+    const empty_list = comptime kvs_list.len == 0;
```
Comment (Vexu) [inline lib/std/comptime_string_map.zig:46]: All the `comptime`s you added are redundant as is the one on line 47.

File: lib/std/comptime_string_map.zig [hunk 2]
```diff
@@ -71,8 +72,8 @@ pub fn ComptimeStringMapWithEql(
         };
         mem.sortUnstableContext(0, sorted_kvs.len, SortContext{ .kvs = &sorted_kvs });
 
-        const min_len = sorted_kvs[0].key.len;
-        const max_len = sorted_kvs[sorted_kvs.len - 1].key.len;
+        const min_len = if (empty_list) -1 else sorted_kvs[0].key.len;
```
Comment (Vexu) [inline lib/std/comptime_string_map.zig:75]: Could you instead break out of the block after line 53 with the empty list?

Comment (rootbeer) [reply lib/std/comptime_string_map.zig]: Ah, yeah I can.  I was originally trying to preserve the type of the result object, but given this is comptime, the empty case can be represented by an empty object.  Fixed!

**Review by rootbeer — COMMENTED** [review]
Thanks for the review.   Fixed everything, I think.
